src/converter/dimacs/DimacsParser.ts

```diff
@@ -0,0 +1,158 @@
+import { Lexer, Token } from "@jlguenego/lexer";
+import { andRule, commentRule, negateRule, orRule, problem, problemTypeKeywords, regexComment, sat, startKeywords } from "./Syntax/DimacsSyntax";
+import { and, or, negate, open, close, variable, parenthesesRule, variableRule, blankRule, regexBlank, variableAssignment } from "./Syntax/CommonSyntax";
+
+// the order is important - tokens are applied from first to last
+const rules = [commentRule, ...startKeywords, ...problemTypeKeywords, ...parenthesesRule, andRule, orRule, negateRule, variableRule, blankRule];
+
+//Inspired by https://github.com/Robbepop/dimacs-parser/blob/master/src/items.rs
+export class DimacsParser {
+    private lex: Lexer = new Lexer(rules);
+    private tokenIterator!: IterableIterator<Token>;
+    private output: string = '';
+    private variables: Set<number> = new Set<number>();
+
+    public parse(text: string): string {
+        this.output = '';
+
+        let tokens: Token[];
+        try {
+            tokens = this.lex.tokenize(text);
+            this.tokenIterator = tokens[Symbol.iterator]()
+        } catch (e: any) {
+            throw new Error(
+                "Couldn't tokenize the input string:\n"
+                + e.message)
+        }
+        if (tokens.length == 0) return this.output;
+
+        //Problem token
+        let token = this.nextToken();
+        if (token.name != problem) new Error("Must start with problem description");
+
+        //Problem type token
+        token = this.nextToken();
+        switch (token.name) {
+            case sat:
+                this.parseSAT()
+                break;
+            default:
+                new Error("Problem " + token.name + " is not supported");
+        }
+
+        //Replace with variable names from comments if possible
+        let aliases = this.getVariableAliases(text);
+        aliases.forEach((alias, num, _) => {
+            //Workaround split, join to replace all
+            this.output = this.output.split(num.toString()).join(alias);
+        });
+
+        return this.output;
+    }
+
+    private getVariableAliases(text: string): Map<number, string> {
+        let aliases = new Map<number, string>();
+
+        let match = text.match(regexComment);
+        if (match == null) return aliases;
+
+        for (let m of match) {
+            let trimmed = m.replace(regexBlank, '');
+
+            //Remove c at beginning
+            trimmed = trimmed.substring(1);
+
+            //Comments must be of type "Number variableAssignment AliasStr"
+            let vars = trimmed.split(variableAssignment);
+            if (vars.length != 2) continue;
+            if (!/\d/.test(vars[0])) continue;
+
+            aliases.set(+vars[0], vars[1]);
+        }
+
+        return aliases;
+    }
```

Comment:
Why doesn't this use the tokenizer? Either approach is fine but it seems complicated to mix tokenization and plain regex

Reply:
I agree that it can be confusing, but in this case I didn't think about using the tokenizer as this requires a strict order of a  number first and then something else. The tokenizer can achieve the same but it would also need to check if the tokens were matched correctly in an order number, var assignment, alias.